asyncpg/protocol/codecs/datetime.pyx

```diff
@@ -232,31 +232,361 @@ cdef timetz_decode(ConnectionSettings settings, FastReadBuffer buf):
+            parts.append("%d day%s" % (self.days, '' if abs(self.days) == 1 else 's'))
+
+        if self.time:
+            time, usecs = divmod(self.time, 1000000)
+            time, secs = divmod(time, 60)
+            time, mins = divmod(time, 60)
+            time, hours = divmod(time, 60)
+            sign = '-' if time < 0 else ''
+            if usecs:
+                parts.append("%s%02d:%02d:%02d.%06d" % (sign, hours, mins, secs, usecs))
+            else:
+                parts.append("%s%02d:%02d:%02d" % (sign, hours, mins, secs))
+
+        return ', '.join(parts)
+
+    def __hash__(self):
+        return hash((self.months, self.days, self.time))
+
+    def __richcmp__(self, other, int op):
+        if isinstance(other, Interval):
+            if op == 2: # ==
+                return (self.months == other.months
+                        and self.days == other.days
+                        and self.time == other.time)
+            elif op == 3: # !=
+                return (self.months != other.months
+                        or self.days != other.days
+                        or self.time != other.time)
+            elif op == 0: # <
+                return ((self.months, self.days, self.time)
+                        <
+                        (other.months, other.days, other.time))
+            elif op == 1: # <=
+                return ((self.months, self.days, self.time)
+                        <=
+                        (other.months, other.days, other.time))
+            elif op == 4: # >
+                return ((self.months, self.days, self.time)
+                        >
+                        (other.months, other.days, other.time))
+            elif op == 5: # >=
+                return ((self.months, self.days, self.time)
+                        >=
+                        (other.months, other.days, other.time))
+        return False
+
+    def __abs__(self):
+        return Interval(abs(self.months),
+                        abs(self.days),
+                        abs(self.time))
+
+    def __neg__(self):
+        return Interval(-self.months,
+                        -self.days,
+                        -self.time)
+
+    def __add__(self, other):
+        if isinstance(self, Interval):
+            if isinstance(other, Interval):
+                return Interval(self.months + other.months,
+                                self.days + other.days,
+                                self.time + other.time)
+
+            if isinstance(other, datetime.date):
+                delta = to_timedelta(self.months, self.days, self.time,
+                                     other.year, other.month, other.day)
+
+                if isinstance(other, datetime.datetime):
+                    result = other
+                else:
+                    result = datetime.datetime(year=other.year, month=other.month, day=other.day)
+
+                return result + delta
+
+        elif isinstance(self, datetime.date):
```

Comment:
How can this branch be ever triggered? `self` should always be an instance of `Interval`, doesn't it?